Add a really simple way to summon new parallel functions later

tizu69 opened the pull request:
I love the idea of coroutine managers like the ones people in the Discord have created, but often times all I need is a simple "hey, create a new coroutine and call it a day" library. However, I hate always having to pull a library for this, to the point that on a server of mine, I have added one to the ROM. This is obviously subpar, as scripts made for the server now no longer work OOB on other CCT worlds or packs.

As such, I propose a backwards-API-compatible way to summon new threads after the fact. This is dead simple - it does not let you kill the thread, makes no assumptions for order (although it may, TODO? I'm not sure how this should be laid out though), and just exists for the purpose of really simple, yet dynamic function parallelism-requiring scripts that would like to stay on stdlib.

meow

tizu69 committed:
Add a really simple way to summon new parallel functions after initial creation (WIP?)

tizu69 committed:
Document said summon functions

tizu69 committed:
Denote that this could be improved

tizu69 committed:
Fix badtype error level

tizu69 committed:
Add tests for postparallelism

scmcgowen reviewed (approved):
Looks good to me

tizu69 commented:
Hi, thanks for your reply :3

> For example, if you're using a `waitForAny`, you cannot `summon` a task which ends without ending the parent tasks as well. Similarly, all child processes must now exit in order for `waitForAll` to exit.
This is something I'm not sure how to solve. Sure, I could give different any/all rules for tasks started afterwards, but what if you don't want that - you want a subtask to end it all? Would I need to add some config object to parallel.waitFor*?

> having `summon` take multiple parameters but only of type `function` feels weird.
summon mimics the arguments of the parallel.waitFor* functions.

> I would rather "summon" each thread individually, especially so since you could then refactor the method to push any inputs to the function, [...]
I have considered this too, but am unsure how I would work this out while still forwarding summon. Right now, summoned functions receive summon, so that they could be part of other files etc etc and you don't need to share your summon function manually. If I allow arguments, you might call, say, rednet.send, which would cause `<recipient>` to be summon function. If I don't default-forward the summon function, you're going to need to create another function either way, to forward summon as needed. This seems like an endless debate - I am happy to implement this once I (or someone else) has found a good solution that fits different usecases - calling external functions, forwarding summon, ...

> I would propose instead a `parallel.run` function that is added to the main coroutines (alongside `rednet.run` and etc). Then, another function `parallel.spawn` that just adds coroutines to the `parallel.run` queue.
This seems like a reasonable idea, although this would mean you pass control over when parallel tasks start to run away from your own scripts and towards the global coroutines, which is fine for most use cases, but :shrug: idk

I have, for now, turned this into a draft PR as I'd like to work out some suggestions before merging this as-is.

SquidDev commented:
Thanks for the PR! Yeah, the coroutine scheduler design space is incredibly tricky, which is partly why I've been avoiding it! Every time I've needed a scheduler, I've ended up writing a new one from scratch, as I always need something slightly different.

The version I posted in https://github.com/cc-tweaked/CC-Tweaked/issues/1734#issuecomment-1971210472 is pretty similar to the `waitForAll` version you've got here, with the main difference being that the `spawn` (or `summon`) function is only passed to the initial function:

```lua
run(function(spawn)
  spawn(function() for i = 1, 3 do print("A", i) sleep(0.5) end end)
  spawn(function(n) for i = 1, n do print("B", i) sleep(0.5) end end, 5)
end)
```

My gut feeling here would be to not support `waitForAny` at all — most of the time when you need a "race" function, it's not dynamic.

You *could* go the route of [`eio`](https://ocaml.org/p/eio/1.3/doc/eio/Eio/Fiber/index.html), and have separate methods to spawn normal and "daemon" coroutines. This is definitely useful IRL, but I'm not sure about here — that's half the problem, lots of things make sense in a fully-featured library like eio or trio, but less so in CC!

```lua
run(function(scope)
  scope:spawn(function() for i = 1, 3 do print("A", i) sleep(0.5) end end)
  scope:spawn_daemon(function(n) for i = 1, n do print("B", i) sleep(0.5) end end, 5)
end)
```

> I would propose instead a `parallel.run` function that is added to the main coroutines (alongside `rednet.run` and etc)

No. Programs shouldn't be able to spawn code that runs outside of their own scope — that way lies madness.

tizu69 commented:
I do have an alternative idea that I came up on the spot with, probably bad though: A `waitForFirst`. It's a `waitForAllOrTheFirstOne`, essentially.
Idea here would be that if you want something that should not be counted towards the "all"ness of waitForAll, you put something like this. this probably comes with tons of drawbacks tho;

```lua
local shouldQuit = false
parallel.waitForAll(..., function()
	parallel.waitForFirst(function()
		while true do
			if shouldQuit then return end -- this will kill all others
			os.pullEvent()
		end
	end, function()
		-- I am unimportant. If I exit, you should not give a fuck.
		-- If I decide not to exit, you should not give a fuck. Kill me.
	end)
end)
```

zyxkad commented:
> Hmm, perhaps implement the parallel.run function as a per-instance system that the user must start in their own program? Much like how many current "thread" libraries work now.

I feel this
```lua
queue = parallel.run(c) -- Runs until `queue.stop()` or no threads are alive
```
is sus. Since parallel.run will block until no threads are alive, that means `queue` will never be assigned before all threads are done? I guess a better way is pass the `queue` as an argument to `c`.

tizu69 commented:
> This case can rewritten with `waitForAll`

@zyxkad This seems like a weird hack, not sure if I like this

fatboychummy commented:
> Since parallel.run will block until no threads are alive, that means `queue` will never be assigned before all threads are done?

That's why `queue` is predeclared at the very top.

zyxkad commented:
> > Since parallel.run will block until no threads are alive, that means `queue` will never be assigned before all threads are done?
> 
> That's why `queue` is predeclared at the very top. This is a common pattern in programming.

Yeah I understand the declaration, but what I'm saying is it will be `nil` till `parallel.run` returned.

fatboychummy commented:
> Yeah I understand the declaration, but what I'm saying is it will be `nil` till `parallel.run` returned.

Ah woops, you're right. Probably a mix of Lupus' idea with mine would work then. You create a queue first, then tell it to run the queue.

tizu69 commented:
couldn't really think of any great solutions without overcomplicating this.

waitForFirst I can do if there's interest, otherwise, I'll mark this ready for review.

SquidDev committed:
Merge branch 'mc-1.20.x' into feat/nested-parallel

SquidDev committed:
Kinda rewrite things

The two parallel implementations are now entirely separate. I don't
think we gained much from trying to share them.

 - Remove spawning support from parallel.waitForAny
 - Prevent spawning from outside the parallel block
 - Clean up spawned tasks
 - Fix spawned functions being called with an event. They can now be
   called with arbitrary arguments.
 - Docs

SquidDev committed:
Fix error levels

SquidDev committed:
Fix/cleaning can_spawn code
